### PR TITLE
Implemented PR feedback from general Scheduler V2 Review

### PR DESCRIPTION
**Pull request by jzila**
Implemented all the minor feedback. Additionally, factored out formation rectification
so only a minimally needed amount of work occurs. This is to address https://github.com/flynn/flynn/pull/1818#discussion_r39878345.

### CODE REVIEW

File: controller/scheduler/discoverd.go [hunk 1]
```diff
@@ -59,6 +59,11 @@ func (d *discoverdWrapper) Register() (bool, error) {
 	outer:
 		for {
 			for leader := range leaders {
+				if leader == nil {
+					// This should never happen, but it has happened, so a nil check is necessary
+					log.Error("Received nil leader event")
```
Comment (lmars) [inline controller/scheduler/discoverd.go]: downcase the `R`

File: controller/scheduler/scheduler.go [hunk 2]
```diff
@@ -722,51 +734,64 @@ func (s *Scheduler) changeFormation(ef *ct.ExpandedFormation) (*Formation, error
...
 			f.Processes = ef.Processes
 		}
 	}
+	s.triggerRectify(f.key())
 	return f, nil
 }
 
+func (s *Scheduler) triggerRectify(key utils.FormationKey) {
+	logger.Info("triggering rectify", "key", key)
+	s.formations.TriggerRectify(key)
```
Comment (lmars) [inline controller/scheduler/scheduler.go]: Does this need to go via `s.formations` and have all the generated case statements? Can we not make `s.rectify` a `chan *Formation`, with `nil` meaning rectify all formations?

File: controller/scheduler/scheduler.go [hunk 3]
```diff
@@ -722,51 +734,64 @@ func (s *Scheduler) changeFormation(ef *ct.ExpandedFormation) (*Formation, error
...
+}
+
 func (s *Scheduler) updateFormation(f *ct.Formation) (*Formation, error) {
 	ef, err := utils.ExpandFormation(s, f)
 	if err != nil {
 		return nil, err
 	}
-	return s.changeFormation(ef)
+	form, err := s.changeFormation(ef)
+	return form, err
```
Comment (lmars) [inline controller/scheduler/scheduler.go]: Is this change necessary?

File: controller/scheduler/scheduler.go (hunk 2 above)
Comment (jzila) [inline controller/scheduler/scheduler.go]: no, I want the advantage of being able to rectify each formation once or not at all.

Comment (jzila) [inline controller/scheduler/scheduler.go]: That is, s.rectify responding to `chan *Formation` has a chance of rectifying multiple times per formation for no reason...thanks to the priority, all events leading up to the "first" rectify will already have happened, so the "first" rectify will be current.

File: controller/scheduler/scheduler.go (hunk 3 above)
Comment (jzila) [inline controller/scheduler/scheduler.go]: No, I originally had it returning more parameters but now it's the same.

File: controller/scheduler/scheduler.go (hunk 2 above)
Comment (lmars) [inline controller/scheduler/scheduler.go]: Ok so I am not really a fan of the reflection and a channel per formation just to avoid the chance of a loop iteration triggering the same rectify multiple times, as the duplicates will effectively be no-ops anyway right? The loop should be fast enough that the chance of having duplicates is low. WDYT?

Comment (lmars) [inline controller/scheduler/scheduler.go]: Thinking about it, another option is to have a buffer wrapper around the rectify channel, and have `s.rectify` receive a "batch" of formations, e.g.:

```go
batch := map[utils.FormationKey]struct{}
var out chan map[utils.FormationKey]struct{}
for {
  select {
  case key := <-in:
    // add key to batch
    batch[key] = struct{}{}
    // switch `out` so we try and send the batch on next loop
    out = s.rectify
  case out <- batch:
    // batch sent, reset
    batch = map[utils.FormationKey]struct{}
    out = nil
  }
}
```

Comment (jzila) [inline controller/scheduler/scheduler.go]: YES! I like this.

Comment (jzila) [inline controller/scheduler/scheduler.go]: Well, I'd do it differently:

```go
func (s *Scheduler) triggerRectify(key utils.FormationKey) {
    s.rectifyBatch[key] = struct{}{}
    select {
        case s.rectify <- struct{}{}:
        default:
    }
}

func (s *Scheduler) HandleRectify() {
    for key := range s.rectifyBatch {
        s.rectifyFormation(key)
    }
}
```
